walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,169 @@
+          /* addBeginningSlot = */ true, /* addEndSlot = */ true);
+          break;
+        } else if (slot.contains(eventTime.end())) { // End of event is part of slot.
+          splitSlotAroundEvent(curSlots, request.getDuration(), slot, eventTime,
+          /* addBeginningSlot = */ false, /* addEndSlot = */ true);
+        } else if (slot.contains(eventTime.start())) { // Beginning of event is part of slot.
+          splitSlotAroundEvent(curSlots, request.getDuration(), slot, eventTime,
+          /* addBeginningSlot = */ true, /* addEndSlot = */ false);
+        } else if (eventTime.contains(slot)) { // Slot is a subset of event.
+          splitSlotAroundEvent(curSlots, request.getDuration(), slot, eventTime,
+          /* addBeginningSlot = */ false, /* addEndSlot = */ false);
+        }
+      }
+    }
+    return curSlots;
+  }
+
+  /*
+   * Split current time slot into 0-2 smaller time slots, given that event at least 
+   * partially overlaps slot. If event is a subset of slot, then both the remaining 
+   * beginning and end slots can potentially be added in. If event's end is part of 
+   * slot, only the difference (end of event - end of slot) can potentially be added in. 
+   * If event's beginning is part of slot, only the difference (beginning of slot - 
+   * beginning of event) can potentially be added in.
+   *
+   * @param curSlots: Current list of time ranges that the meeting request can take 
+   * place within. This is modified over the course of the function.
+   * @param reqDuration: duration (in minutes) of the requested meeting.
+   * @param slot: current member of curSlots to be sliced into smaller fragments (before 
+   * and after)
+   * @param eventTime: time range that the event currently be considered takes place in.
+   * @param addBeginningSlot: whether the time range (beginning of slot - beginning of
+   * event) should be considered as a valid time slot for the requested meeting.
+   * @param addEndSlot: whether the time range (end of event - end of slot) should
+   * be considered as a valid time slot for the requested meeting.
+   */
+  private void splitSlotAroundEvent(ArrayList<TimeRange> curSlots, long reqDuration,
+    TimeRange slot, TimeRange eventTime, boolean addBeginningSlot, boolean addEndSlot) {
+
+    TimeRange before = TimeRange.fromStartEnd(slot.start(), eventTime.start(), false);
+    TimeRange after = TimeRange.fromStartEnd(eventTime.end(), slot.end() - 1, true);
+
+    if (addBeginningSlot && before.duration() >= reqDuration) {
+      curSlots.add(before);
+    }
+    if (addEndSlot && after.duration() >= reqDuration) {
+      curSlots.add(after);
+    }
+    curSlots.remove(slot);
+  }
+
+  /*
+   * @groupA, groupB: generalized lists of strings to be compared.
+   * @return whether there is any overlap between groupA and groupB (i.e. a String
+   * appears in both groupA and groupB.
+   */
+  private boolean hasNothingInCommon(Collection<String> groupA, Collection<String> groupB) {
```

Comment:
So this is basically checking if the intersection of two sets is empty.  You could look into using retainAll instead of doing this yourself.

Reply:
It seems like using retainAll would modify the Collection objects themselves -- I replaced this method with a call to Collections.disjoint() instead.